libcontainer/nsenter/nsexec.c

```diff
@@ -311,6 +307,220 @@ static int child_func(void *arg)
+	 * of machines.
+	 *
+	 * (To be honest, maybe it's better to just hope Go doesn't notice any
+	 * issues with glibc rather than trying to hack internal glibc
+	 * structures to make them "work" with Go. But it seems we need to do
+	 * this...)
+	 */
+
+	if (tid_addr)
+		write_log(DEBUG,
+			  "clone: find_glibc_tls_tid_address: PR_GET_TID_ADDRESS is not the cached tid field (*%p [%d] != %d, pthread_self=%p)",
+			  tid_addr, *tid_addr, actual_tid, (void *)pthread_self());
+	else
+		write_log(DEBUG, "clone: find_glibc_tls_tid_address: PR_GET_TID_ADDRESS failed: %m");
+	write_log(WARNING,
+		  "clone: find_glibc_tls_tid_address: falling back to scanning pthread_self(). Please use a kernel with CONFIG_CHECKPOINT_RESTORE=y.");
+
+	/*
+	 * These offsets are based on glibc 2.39, but the layout of struct
+	 * pthread (at least up to the tid field) has been stable for several
+	 * decades. The cached pid (from pre-2.25 glibc) was stored after the
+	 * tid field, so even on ancient glibc versions it's "safe" for us to
+	 * do this.
+	 *
+	 * The structure layouts can be found in <sysdeps/.../ntpl/tls.h>. Only a
+	 * few architectures have tcbhead_t in the pthread header (TLS_TCB_AT_TP),
+	 * so we only need to define the structure for those architectures.
+	 */
+
+#  if defined(__x86_64__) || defined(__i386__) || defined(__s390__) || defined(__s390x__) || defined(__sparc__)
+#    define TLS_TCB_AT_TP 1
+#  else
+#    define TLS_TCB_AT_TP 0
+#  endif
+
+#  if TLS_TCB_AT_TP
+	struct tcbhead_t {
+#    if defined(__x86_64__)
+		void *tcb, *dtv, *self;
+		int multiple_threads, gscope_flag;
+		uintptr_t sysinfo, stack_guard, pointer_guard;
+		unsigned long int unused_vgetcpu_cache[2];
+		unsigned int feature_1;
+		int __glibc_unused1;
+		void *__private_tm[4];
+		void *__private_ss;
+		unsigned long long int ssp_base;
+		struct {
+			int i[4];
+		} __glibc_unused[8][4] __attribute__((aligned(32)));
+		void *__padding[8];
+#    elif defined(__i386__)
+		void *tcb, *dtv, *self;
+		int multiple_threads;
+		uintptr_t sysinfo, stack_guard, pointer_guard;
+		int gscope_flag;
+		unsigned int feature_1;
+		void *__private_tm[3];
+		void *__private_ss;
+		unsigned long ssp_base;
+#    elif defined(__s390__) || defined(__s390x__)
+		void *tcb, *dtv, *self;
+		int multiple_threads;
+		uintptr_t sysinfo;
+		uintptr_t stack_guard;
+		int gscope_flag;
+		int __glibc_reserved1;
+		void *__private_ss;
+#    elif defined(__sparc__)
+		void *tcb, *dtv, *self;
+		int multiple_threads;
+#      if __WORDSIZE == 64
+		int gscope_flag;
+#      endif
+		uintptr_t sysinfo;
+		uintptr_t stack_guard;
+		uintptr_t pointer_guard;
+#      if __WORDSIZE != 64
+		int gscope_flag;
+#      endif
+#    endif
+	};
+#  endif
+       /* TLS_TCB_AT_TP */
+
+	/* Based on <ntpl/descr.h>. */
+	struct __glibc_pthread {
+		union {
+#  if TLS_TCB_AT_TP		/* !TLS_DIV_AT_TP */
+			struct tcbhead_t header;
+#  else
+			struct {
+				int multiple_threads, gscope_flag;
+			} header:
```

Comment:
Do we want ';' here?
```suggestion
			} header;
```

Reply:
typo